blockchain/chain_sync/src/tipset_syncer.rs

```diff
@@ -1398,159 +1320,53 @@ async fn validate_block<
-        let v_prev_beacon = Arc::clone(&prev_beacon);
-        validations.push(task::spawn(async move {
-            v_block
-                .header()
-                .validate_block_drand(
-                    win_p_nv,
-                    beacon_schedule.as_ref(),
-                    parent_epoch,
-                    &v_prev_beacon,
-                )
-                .await
-                .map_err(|e| {
-                    TipsetRangeSyncerError::Validation(format!(
-                        "Failed to validate blocks random beacon values: {}",
-                        e
-                    ))
-                })
-        }));
-    }
-
-    // Ticket election proof validations
-    let v_block = Arc::clone(&block);
-    let v_prev_beacon = Arc::clone(&prev_beacon);
-    validations.push(task::spawn_blocking(move || {
-        let header = v_block.header();
-        let mut miner_address_buf = header.miner_address().marshal_cbor()?;
-
-        if header.epoch() > smoke_height {
-            let vrf_proof = base_tipset
-                .min_ticket()
-                .ok_or(TipsetRangeSyncerError::TipsetWithoutTicket)?
-                .vrfproof
-                .as_bytes();
-            miner_address_buf.extend_from_slice(vrf_proof);
-        }
-
-        let beacon_base = header.beacon_entries().last().unwrap_or(&v_prev_beacon);
-
-        let vrf_base = state_manager::chain_rand::draw_randomness(
-            beacon_base.data(),
-            DomainSeparationTag::TicketProduction as i64,
-            header.epoch() - TICKET_RANDOMNESS_LOOKBACK,
-            &miner_address_buf,
-        )
-        .map_err(|e| TipsetRangeSyncerError::DrawingChainRandomness(e.to_string()))?;
-
-        verify_election_post_vrf(
-            &work_addr,
-            &vrf_base,
-            // Safe to unwrap here because of block sanity checks
-            header.ticket().as_ref().unwrap().vrfproof.as_bytes(),
-        )?;
-
-        Ok(())
-    }));
-
-    // Winning PoSt proof validation
     let v_block = block.clone();
-    let v_prev_beacon = Arc::clone(&prev_beacon);
-    validations.push(task::spawn_blocking(move || {
-        verify_winning_post_proof::<_, V>(
-            &state_manager,
-            win_p_nv,
-            v_block.header(),
-            &v_prev_beacon,
-            &lookback_state,
-        )?;
-        Ok(())
+    validations.push(task::spawn(async move {
+        consensus
+            .validate_block(state_manager, v_block)
+            .map_err(|errs| {
+                // NOTE: Concatentating errors here means the wrapper type of error
+                // never surfaces, yet we always pay the cost of the generic argument.
+                // But there's no reason `validate_block` couldn't return a list of all
+                // errors instead of a single one that has all the error messages,
+                // removing the caller's ability to distinguish between them.
+                let mut msgs = vec![];
```

Comment:
does it make sense to `reserve` `errs.len()` upfront here?

Reply:
Thanks for mentioning this. I just noticed that not 10 lines further down I'm using map and collect instead of looping, so it wasn't very consistent. I moved both into a helper function where `with_capacity` is used.

Reply:
I agree on the one liners - I went back to the collect method, although technically it's 4 lines because of the formatter :man_shrugging: